### ISSUE TITLE
128-bit integers marked as having an unknown stable ABI

### ISSUE THREAD

**Issue by haraldh** [rust-lang/rust#78473]
Introduced by 549dd105273a5141003b6054c6f496f62e059cfb 128-bit integers used in e.g. `extern "sysv64"` produce the warning:
```
warning: `extern` fn uses type `u128`, which is not FFI-safe
   --> src/xxx.rs:126:6
    |
126 | ) -> u128 {
    |      ^^^^ not FFI-safe
    |
    = note: `#[warn(improper_ctypes_definitions)]` on by default
    = note: 128-bit integers don't currently have a known stable ABI

warning: 1 warning emitted
```

I dispute this statement by pointing to the now de facto source of the x86-64 System V ABI: 

* https://gitlab.com/x86-psABIs/x86-64-ABI/-/blob/master/x86-64-ABI/low-level-sys-info.tex#L552-560
* https://gitlab.com/x86-psABIs/x86-64-ABI/-/blob/master/x86-64-ABI/low-level-sys-info.tex#L767-768

**Comment by est31** [comment]
Good point actually. Allowing it for `extern "C"` for `sysv64` using targets would probably be a bad idea due to the current lack of a portability lint (#41619), but allowing it for `extern "sysv64"` should be doable I think.

cc @nagisa

**Comment by est31** [comment]
Made a PR to allow it on sysv64: #78481

**Comment by workingjubilee** [comment]
Making some notes here for the next swing at a generic integers RFC:
- `_BitInt(N)` is upcoming in C23
- Current spec for SysV x86-64 `_BitInt(N) where N > 64` is as-if eightbyte-aligned structs
- Thus, `__int128` and `_BitInt(128)` are **ABI-incompatible**
- If we RFC something like `int<const N>` we may want to decide:
  - `type i128 = int<128>;`
  - `int<N>` is `repr(C)`-compatible as `_BitInt(N)`
- That would make this pair incorrect:
```rust
extern "C" fn(a: i128, b: i128) -> i128
```
```c
__int128 by_value_int128s(__int128 a, __int128 b);
```
- Deciding otherwise means we can't call this C23 function:
```c
_BitInt(128) by_value_big_bitints(_BitInt(128) a, _BitInt(128) b);
```

I have opened [an issue regarding this at the x86-64-ABI gitlab](https://gitlab.com/x86-psABIs/x86-64-ABI/-/issues/11). However they may have good reasons for their current spec, so it is not guaranteed they will consider this a spec bug.

**Comment by tgross35** [comment]
I think the links in the original issue are dead, unless this was actually the intended highlight range

![image](https://github.com/rust-lang/rust/assets/13724985/a308ab29-885b-4e65-a638-8e4c0e4f8127)

(GL changes the URL if you hit the "permalink" button top right, not like GH where you can copy a permalink directly)

**Comment by tgross35** [comment]
Did a bit of sleuthing, I believe these are the two relevant sections (based on the commit at time of OP):

https://gitlab.com/x86-psABIs/x86-64-ABI/-/blob/ca48acd297294172349d4057fb0f04016c5733bd/x86-64-ABI/low-level-sys-info.tex#L591-602

https://gitlab.com/x86-psABIs/x86-64-ABI/-/blob/ca48acd297294172349d4057fb0f04016c5733bd/x86-64-ABI/low-level-sys-info.tex#L814-815
